### PR TITLE
Add UDF support to c++ API

### PR DESCRIPTION
**Pull request by acquamarin**
Implement UDF support to C++ API.

### CODE REVIEW

File: src/catalog/catalog.cpp [hunk 1]
```diff
@@ -463,5 +463,11 @@ std::unordered_set<RelTableSchema*> Catalog::getAllRelTableSchemasContainBoundTa
     return relTableSchemas;
 }
 
+void Catalog::addVectorFunction(
```
Comment (andyfengHKU) [inline src/catalog/catalog.cpp:466]: open an issue about renaming vector function to scalar function

File: src/include/function/built_in_vector_functions.h [hunk 2]
```diff
@@ -27,6 +27,14 @@ class BuiltInVectorFunctions {
     static uint32_t getCastCost(
         common::LogicalTypeID inputTypeID, common::LogicalTypeID targetTypeID);
 
+    inline void addFunction(std::string name, function::vector_function_definitions definitions) {
```
Comment (andyfengHKU) [inline src/include/function/built_in_vector_functions.h:30]: move to cpp. I start realizing it's better to move more code into cpp file if the header being included in many places. So we don't need to recompile a lot of files when the design changes

File: src/include/main/connection.h [hunk 3]
```diff
@@ -202,6 +213,8 @@ class Connection {
         return queryResultWithError(exceptionMessage);
     }
 
+    void addVectorFunction(std::string name, function::vector_function_definitions definitions);
```
Comment (andyfengHKU) [inline src/include/main/connection.h:216]: addScalarFunction

File: test/main/udf_test.cpp [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        *conn->query("MATCH (p:person) return strDoubleLen(p.fName)"));
+    auto expectedResult = std::vector<std::string>{"10", "6", "10", "6", "18", "12", "8", "98"};
+    sortAndCheckTestResults(actualResult, expectedResult);
+}
+
+int64_t addSecondParamTwice(int16_t x, int32_t y) {
+    return x + y + y;
+}
+
+TEST_F(ApiTest, BinaryUDFInt) {
```
Comment (andyfengHKU) [inline test/main/udf_test.cpp:60]: Add a test where input can be flat and unflattering

**Review by andyfengHKU — APPROVED** [review]

File: src/catalog/catalog.cpp (hunk 1 above)
Comment (acquamarin) [reply src/catalog/catalog.cpp:466]: #1809

**Pull request merged**